src/recipe/Baked.Recipe.Service.Application/DataAccess/DataAccessLayer.cs

```diff
@@ -16,75 +16,69 @@ public class DataAccessLayer : LayerBase<AddServices, PostBuild>
-                .ExposeConfiguration(c => c.SetInterceptor(new DelegatedInterceptor(sp, _interceptorConfiguration)))
-                .Mappings(m => m.AutoMappings.Add(_autoPersistenceModel));
-
-            if (_persistenceConfiguration.AutoUpdateSchema)
-            {
-                builder.ExposeConfiguration(c => new SchemaUpdate(c).Execute(false, true));
-            }
-
-            return builder.BuildConfiguration();
-        });
-
-        services.AddSingleton(sp => sp.GetRequiredService<NHConfiguration>().BuildSessionFactory());
-        services.AddSingleton<Func<ISession>>(sp => () => sp.GetRequiredServiceUsingRequestServices<ISession>());
-
-        services.AddScoped(sp =>
-        {
-            var result = sp.GetRequiredService<ISessionFactory>().OpenSession();
-
-            if (_persistenceConfiguration.AutoExportSchema)
-            {
-                if (!_exported)
-                {
-                    lock (_exportedLock)
-                    {
-                        if (!_exported)
-                        {
-                            var export = new SchemaExport(sp.GetRequiredService<NHConfiguration>());
-
-                            export.Execute(false, true, false, result.Connection, null);
-                            _exported = true;
-                        }
-                    }
-                }
-            }
-
-            return result;
-        });
 
         return phase.CreateContextBuilder()
+            .Add(_fluentConfiguration)
             .Add(_persistenceConfiguration)
             .Add(_interceptorConfiguration)
             .Add(_automappingConfiguration)
             .Add(_autoPersistenceModel)
+            .OnDispose(() =>
+            {
+                _fluentConfiguration.Database(_persistenceConfiguration.Configurer);
+                _fluentConfiguration.Mappings(m => m.AutoMappings.Add(_autoPersistenceModel));
+
+                services.AddSingleton(sp => _fluentConfiguration.BuildConfiguration());
+                services.AddSingleton(sp => sp.GetRequiredService<NHConfiguration>().BuildSessionFactory());
+                services.AddScoped(sp => sp.GetRequiredService<ISessionFactory>().OpenSession());
+                services.AddSingleton<Func<ISession>>(sp => () => sp.GetRequiredServiceUsingRequestServices<ISession>());
+            })
             .Build();
     }
 
     protected override PhaseContext GetContext(PostBuild phase)
     {
         var sp = Context.GetServiceProvider();
+
         NHibernateLogger.SetLoggersFactory(sp.GetRequiredService<INHibernateLoggerFactory>());
+        sp.GetRequiredService<NHConfiguration>().SetInterceptor(new DelegatedInterceptor(sp, _interceptorConfiguration));
+
+        if (_persistenceConfiguration.AutoExportSchema)
+        {
+            if (Context.Has<IServiceScope>())
+            {
+                ExportSchema(Context.GetServiceScope().ServiceProvider);
+            }
```

Comment:
now that we decided to use multiple scopes even during the test run, you can remove `IServiceScope` from context, and remove this if block.

Reply:
you may inline `ExportSchema` method as well, since there it will be used only once